File: mark_wutka+haskell/digitrec_array.hs
import System.Environment (getArgs)
import Data.List
import Data.Array.Unboxed
import Data.Ord (compare)

-- Array-version of digit recognizer
-- In theory it should be much faster to process the samples as arrays
-- because they are stored in a contiguous block of memory and modern
-- processors are optimized to pre-fetch contiguous locations.
-- In practice it only runs twice as fast as the list version

-- Compile with: ghc -O2 digitrec_array.hs
-- Run with: digitrec_array testsamplefile trainingsamplefile
-- Example run: digitrec_array ../test-sample.csv ../training-sample.csv

replaceComma ',' = ' '
replaceComma c = c

-- Parses a comma-separated string of numbers into an array of ints
-- Haskell words function only splits on space so this function changes
-- commas to spaces before calling words. Then it maps the read function
-- over the list of strings returned by words, then creates an array from that
parseLine :: String -> UArray Int Int
parseLine s =
    listArray (0, (length lineData) - 1) lineData
       where
         lineData = (map read $ words (map replaceComma s))

-- Reads the file content, uses lines to split it into a list of strings,
-- then maps parseLine over the list of lines
loadData filename = do
    filedata <- readFile filename
    return (map parseLine (tail (lines filedata)))

-- Compares two lists of pixels by summing the squares of the differences
-- Returns a pair containing the test sample digit and its score
compareSamples :: UArray Int Int -> UArray Int Int -> (Int,Int)
compareSamples trainingSample testSample =
  (testSample ! 0,
    sum $ map comparePixels (tail (indices trainingSample)))
      where
        comparePixels i = sqr((trainingSample ! i) - (testSample ! i))
        sqr x = x * x

-- Finds the closes match by looking for the one with the minimum score
recognizeSample :: UArray Int Int -> [UArray Int Int] -> (Int,Int)
recognizeSample trainingSample testSamples =
  minimumBy compareScores (map (compareSamples trainingSample) testSamples)
    where
      compareScores (d1,s1) (d2,s2) = compare s1 s2

-- Returns true if the recognizer recognized the correct digit
sampleMatches :: [UArray Int Int] -> UArray Int Int -> Bool
sampleMatches testSamples trainingSample =
    trainingSample ! 0 == fst (recognizeSample trainingSample testSamples)

main = do
  -- Get the command line argumentsw
  argv <- getArgs
  let [testfile,trainingfile] = take 2 argv

  -- Load the data
  testSample <- loadData testfile
  trainingSample <- loadData trainingfile

-- Count the number of matches
  let numMatched = length $ filter (sampleMatches trainingSample) testSample
  let successRate = (fromIntegral numMatched) / (fromIntegral $ length testSample)
  putStrLn $ show successRate

File: mark_wutka+ocaml/digitrec_array.ml
(* OCaml Array version of Digit Recognizer (lines of numbers are
 * stored in an array instead of a list for faster access
 * To compile:
 *   ocamlopt -o digitrec_array str.cmxa digitrec_array.ml
 * To run:
 *  digitrec_array ../test-sample.csv ../training-sample.csv
 *
 * On my system it runs in about 10.4 seconds.
 *)

(* Splits a comma-separated list of numbers into a pair containing the first
 * number in the list (the match digit) and an array containing the rest of the
 * numbers *)
let parseLine line =
  let values = List.map int_of_string (Str.split (Str.regexp "[,;]") line)
  in
    (List.hd values, Array.of_list (List.tl values))

let parseFile lines =
  List.map parseLine lines;;

let make_stream filename =
  let channel = open_in filename in
    Stream.from (fun _ ->
      try Some (input_line channel) with End_of_file -> None);;

(* Reads a file and returns its contents as a list of lines *)
let lines filename =
  let rec readLines channel lineList =
    try readLines channel ((input_line channel) :: lineList)
    with End_of_file -> List.rev lineList
  in
    readLines (open_in filename) [];;

let arrayZipFold f init a1 a2 =
  let rec iterator i accum =
    if (i >= Array.length a1) || (i >= Array.length a2) then
      accum
    else
      iterator (i+1) (f (Array.get a1 i) (Array.get a2 i) accum)
  in
    iterator 0 init;;

(* Compares two arrays of pixels and returns the sum of the squares of the differences *)
let compareSamples (testCh,testSample) (trainingCh,trainingSample) =
  let sqr x = x * x in
  let comparePixels p1 p2 = sqr (p1 - p2) in
  let pixelSum p1 p2 s = s + comparePixels p1 p2
  in
  (trainingCh,
    arrayZipFold pixelSum 0 testSample trainingSample);;

(* Finds the closest match by looking for the pair with the lowest difference score *)
let recognizeSample testSample trainingSamples =
  let minimumScore (d1,s1) (d2,s2) = if s1 < s2 then (d1,s1) else (d2,s2)
in
  List.fold_left minimumScore (0-1,999999999) (List.map (compareSamples testSample) trainingSamples);;

(* Returns true if the recognizer recognized the correct digit *)
let sampleMatches trainingSamples (testCh,testSample) =
  testCh = fst (recognizeSample (testCh,testSample) trainingSamples);;

let runme () =
  let testFile = Sys.argv.(1) in
  let trainingFile = Sys.argv.(2) in
  let testSample = parseFile (List.tl (lines testFile)) in
  let trainingSample = parseFile (List.tl (lines trainingFile)) in
  let numMatched = List.length (List.filter (sampleMatches trainingSample) testSample) in
  let numSamples = List.length testSample in
  let successRate = (float_of_int numMatched) /. (float_of_int numSamples)
in
  print_float successRate;
  print_newline();;

runme();;
